tools/generate_definitions.py

```diff
@@ -0,0 +1,375 @@
+        "type": "Amount"
+      }
+    ],"""
+)
+
+# Parse STypes
+# Example line:
+# STYPE(STI_UINT32, 2)    \
+type_hits = re.findall(
+    r"^ *STYPE\(STI_([^ ]*?) *, *([0-9-]+) *\) *\\?$", sfield_h, re.MULTILINE
+)
+if len(type_hits) == 0:
+    type_hits = re.findall(
+        r"^ *STI_([^ ]*?) *= *([0-9-]+) *,?$", sfield_h, re.MULTILINE
+    )
+# name-to-value map - needed for SField processing
+type_map = {x[0]: x[1] for x in type_hits}
+
+
+def _is_vl_encoded(t: str) -> str:
+    if t == "VL" or t == "ACCOUNT" or t == "VECTOR256":
+        return "true"
+    return "false"
+
+
+def _is_serialized(t: str, name: str) -> str:
+    if t == "LEDGERENTRY" or t == "TRANSACTION" or t == "VALIDATION" or t == "METADATA":
+        return "false"
+    if name == "hash" or name == "index":
+        return "false"
+    return "true"
+
+
+def _is_signing_field(t: str, not_signing_field: str) -> str:
+    if not_signing_field == "notSigning":
+        return "false"
+    if t == "LEDGERENTRY" or t == "TRANSACTION" or t == "VALIDATION" or t == "METADATA":
+        return "false"
+    return "true"
+
+
+# Parse SField.cpp for all the SFields and their serialization info
+# Example lines:
+# TYPED_SFIELD(sfFee, AMOUNT, 8)
+# UNTYPED_SFIELD(sfSigners,  ARRAY, 3, SField::sMD_Default, SField::notSigning)
+sfield_hits = re.findall(
+    r"^ *[A-Z]*TYPED_SFIELD *\( *sf([^,\n]*),[ \n]*([^, \n]+)[ \n]*,[ \n]*"
+    r"([0-9]+)(,.*?(notSigning))?",
+    sfield_macro_file,
+    re.MULTILINE,
+)
+sfield_hits += [
+    ("hash", "UINT256", "257", "", "notSigning"),
+    ("index", "UINT256", "258", "", "notSigning"),
+]
+sfield_hits.sort(key=lambda x: int(type_map[x[1]]) * 2**16 + int(x[2]))
+for x in range(len(sfield_hits)):
+    _add_line("    [")
+    _add_line('      "' + sfield_hits[x][0] + '",')
+    _add_line("      {")
+    _add_line(
+        '        "isSerialized": '
+        + _is_serialized(sfield_hits[x][1], sfield_hits[x][0])
+        + ","
+    )
+    _add_line(
+        '        "isSigningField": '
+        + _is_signing_field(sfield_hits[x][1], sfield_hits[x][4])
+        + ","
+    )
+    _add_line('        "isVLEncoded": ' + _is_vl_encoded(sfield_hits[x][1]) + ",")
+    _add_line('        "nth": ' + sfield_hits[x][2] + ",")
+    _add_line('        "type": "' + _translate(sfield_hits[x][1]) + '"')
+    _add_line("      }")
+    _add_line("    ]" + ("," if x < len(sfield_hits) - 1 else ""))
+
+_add_line("  ],")
+
+########################################################################
+#  Ledger entry type processing
+########################################################################
+_add_line('  "LEDGER_ENTRY_TYPES": {')
+
+
+def _unhex(x: str) -> str:
+    if x[0:2] == "0x":
+        return str(int(x, 16))
+    return x
+
+
+# Parse ledger entries
+# Example line:
+# LEDGER_ENTRY(ltNFTOKEN_OFFER, 0x0037, NFTokenOffer, nft_offer, ({
+lt_hits = re.findall(
+    r"^ *LEDGER_ENTRY[A-Z_]*\(lt[A-Z_]+ *, *([x0-9a-f]+) *, *([^,]+), *([^,]+), \({$",
```

Comment:
This regex is sensitive to a particular way of writing hex-numbers. It doesn't capture `0Xab` or `0xAB`.

Today's version of rippled code does not cause a problem. However, future versions of rippled code could break this regex.

Reply:
I'm referring to the 0'th capture group

Reply:
fixed